### PR TITLE
Locksmith now builds with TypeScript

### PR DESCRIPTION
**Pull request by cnasc**
# Description

Following on #2050, the experiment to build Locksmith with TypeScript, this PR introduces the minimum configuration necessary to achieve that state.

Change summary:
1. Install dependent packages
2. Add `tsc` script to `package.json`
2. Generate and customize `tsconfig.json`
3. Modify `dev` script in `package.json` to allow the app to automatically build and restart with TypeScript
4. Update `.gitignore` and `.eslintignore` to avoid build artifacts
5. Upgrade Jest
  a. `ts-jest` requires v24 of `jest`
  b. upgrading to `jest@24.1` revealed that we had a malformed test, which was fixed
  c. upgrading to a new major version of `jest` changed the way code coverage is calculated, so the thresholds had to be adjusted for the build to pass.

As with #2071, no actual TypeScript is included in this PR. This only allows the ability to use TypeScript in the future.

# Issues

Fixes #2050 

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [x] This PR only contains configuration changes (package.json, etc.) **[and a test, but the config changes couldn't pass without it]**
  - [ ] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: locksmith/package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "main": "server.js",
   "scripts": {
     "predev": "NODE_ENV=development sequelize db:migrate",
-    "dev": "nodemon server.js",
+    "dev": "tsc-watch --onSuccess 'node build/server.js' --onFailure 'echo Beep! Compilation Failed'",
```
Comment (julien51) [inline locksmith/package.json]: :beep:

**Comment by cnasc** [comment]
@akeem and I are going to go over this today to make sure there are no negative implications for the deploy process, and once that checks out I will merge.

File: locksmith/scripts/deploy-elasticbeanstalk.sh [hunk 2]
```diff
@@ -22,10 +22,10 @@ function package_application()
 {
     local application=${1}
     local artifact_location=${2}
-
+    
     mkdir builds
-    pushd ./${application}
-    zip ../${artifact_location} -r * .[^.]* --exclude=*node_modules*
+    pushd ./${application}/build
+    zip -r ../${artifact_location} . ../package.json ../package-lock.json
```
Comment (akeem) [inline locksmith/scripts/deploy-elasticbeanstalk.sh]: For the time being please add shrinkwrap as well here

File: locksmith/scripts/deploy-elasticbeanstalk.sh [hunk 3]
```diff
@@ -22,10 +22,10 @@ function package_application()
 {
     local application=${1}
     local artifact_location=${2}
-
+
```
Comment (akeem) [inline locksmith/scripts/deploy-elasticbeanstalk.sh]: nit: we should probably remove this

**Comment by cnasc** [comment]
:shipit:

**Comment by cnasc** [comment]
Hm, this is stuck in CI-land so I'm going to close and reopen after I fix the merge conflict.

**Comment by julien51** [comment]
Oups! I see this PR re-introduced the `app.test.js` file which had been deleted in the meantime...

**Pull request merged**